homeassistant/components/logbook/__init__.py

```diff
@@ -442,7 +473,9 @@ def _keep_event(hass, event, entities_filter):
     else:
         event_data = event.data
         entity_id = event_data.get(ATTR_ENTITY_ID)
-        if entity_id is None:
+        if entity_id:
+            context_entity_id_map.setdefault(event.context_id, entity_id)
```

Comment:
I'm not sure if showing attribution to an entity is correct.

These are the things that can initiate entity state changes from inside our system:

 - call service (directly via API/UI)
 - script execute (directly via API/UI)
 - automation triggered (based on event happening)

By looking at entity ID we correctly capture script and automation events. However, for a service call event it might not be enough. A service call can also target multiple entities or an area. 

I think that we should use an explicit implementation where we test for the different types of events and attribute them accordingly.

Reply:
Looks like we have

`call_service`
`script_started`
`automation_triggered`
`state_changed` (maybe this triggers templates)

I tried add another level to the `dict` and it got really slow.  Will try another approach.

Reply:
I've covered them all at this point.  Now working on making if faster

Reply:
Total performance hit is around 3-5% so not bad.